__internal__/modelHookReducer.ts

```diff
@@ -0,0 +1,112 @@
+import {Model, ModelUtils, ModelValueList} from "../hooks"
+import {isType} from "../utility/assertions"
+import {len} from "../utility/collection"
+import {ParseableInput, parseInputValue} from "../utility/dom"
+import {Action, createAction, isSyntheticEvent} from "../utility/react"
+import {cast} from "../utility/string"
+import {call} from "../utility/function"
+
+interface State {
+    // Attempts to prevent unnecessary validation.
+    hasChanged: boolean
+    model: Model<any>
+    utils: ModelUtils
+}
+
+type Actions =
+    | Action<ModelValueList<any>, "change">
+    | Action<undefined, "validate">
+
+/**
+ * This is the reducer for useModel hook.
+ */
+export function modelHookReducer(state: State, action: Actions): State {
+    const {utils} = state
+
+    switch (action.type) {
+        case "change": {
+            const values = action.value
+            // Bail out of rendering if there are no values to update.
+            if (!len(values)) {
+                return state
+            }
+
+            const model = {...state.model}
+            for (const name in values) {
+                const entry = model[name]
```

Comment:
Model entries should not be mutated.

```suggestion
                const entry = {...model[name]}
```